component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java

```diff
@@ -110,293 +110,477 @@ public AuthenticatorFlowStatus process(HttpServletRequest request,
-        } catch (SMSOTPException e) {
-            throw new AuthenticationFailedException("Failed to get the parameters from authentication xml fie", e);
+            throw new AuthenticationFailedException("Authentication Failed: An IOException was caught. ", e);
         }
     }
 
+    /**
+     * Redirect the user to mobile number request page.
+     *
+     * @param response    the HttpServletResponse
+     * @param context     the AuthenticationContext
+     * @param queryParams the queryParams
+     * @throws AuthenticationFailedException
+     */
+    private void redirectToMobileNoReqPage(HttpServletResponse response, AuthenticationContext context,
+                                           String queryParams) throws AuthenticationFailedException {
+        boolean isEnableMobileNoUpdate = SMSOTPUtils.isEnableMobileNoUpdate(context, getName());
+        if (isEnableMobileNoUpdate) {
+            String loginPage = SMSOTPUtils.getMobileNumberRequestPage(context, getName());
+            try {
+                response.sendRedirect(loginPage + "?" + queryParams + SMSOTPConstants.AUTHENTICATORS + getName());
+            } catch (IOException e) {
+                throw new AuthenticationFailedException("Authentication failed!. An IOException was caught. ", e);
+            }
+        } else {
+            throw new AuthenticationFailedException("Authentication failed!. Update mobile no in your profile.");
+        }
+    }
 
     /**
      * Process the response of the SMSOTP end-point.
+     *
+     * @param request  the HttpServletRequest
+     * @param response the HttpServletResponse
+     * @param context  the AuthenticationContext
+     * @throws AuthenticationFailedException
      */
     @Override
     protected void processAuthenticationResponse(HttpServletRequest request, HttpServletResponse response,
                                                  AuthenticationContext context) throws AuthenticationFailedException {
-        Map<String, String> smsOTPParameters = getAuthenticatorConfig().getParameterMap();
         String userToken = request.getParameter(SMSOTPConstants.CODE);
         String contextToken = (String) context.getProperty(SMSOTPConstants.OTP_TOKEN);
+        AuthenticatedUser authenticatedUser = (AuthenticatedUser) context.getProperty(SMSOTPConstants.AUTHENTICATED_USER);
+        if (StringUtils.isEmpty(request.getParameter(SMSOTPConstants.CODE))) {
+            throw new InvalidCredentialsException("Code cannot not be null");
+        }
+        if (Boolean.parseBoolean(request.getParameter(SMSOTPConstants.RESEND))) {
+            if (log.isDebugEnabled()) {
+                log.debug("Retrying to resend the OTP");
+            }
+            throw new InvalidCredentialsException("Retrying to resend the OTP");
+        }
+        if (userToken.equals(contextToken)) {
+            context.setSubject(authenticatedUser);
+        } else if (SMSOTPUtils.getBackupCode(context, getName()).equals("false")) {
```

Comment:
Checking with backup code also configurable. If that property false, then no need to check with backup codes, it will allowed to retry with another code.